// @sandstreamdev/react-swipeable-list/dist/react-swipeable-list.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('react'), require('prop-types')) :
  typeof define === 'function' && define.amd ? define(['exports', 'react', 'prop-types'], factory) :
  (global = global || self, factory(global['@sandstreamdev/react-swipeable-list'] = {}, global.React, global.PropTypes));
}(this, (function (exports, React, PropTypes) { 'use strict';

  var React__default = 'default' in React ? React['default'] : React;
  PropTypes = PropTypes && PropTypes.hasOwnProperty('default') ? PropTypes['default'] : PropTypes;

  var styles = {"swipeableList":"SwipeableList_swipeableList__3dl6y"};

  var SwipeableList = function SwipeableList(_ref) {
    var children = _ref.children,
        scrollStartThreshold = _ref.scrollStartThreshold,
        swipeStartThreshold = _ref.swipeStartThreshold,
        threshold = _ref.threshold;
    return React__default.createElement("div", {
      className: styles.swipeableList,
      "data-testid": "list-wrapper"
    }, React__default.Children.map(children, function (child) {
      return React__default.cloneElement(child, {
        scrollStartThreshold,
        swipeStartThreshold,
        threshold
      });
    }));
  };

  SwipeableList.propTypes = {
    children: PropTypes.node,
    scrollStartThreshold: PropTypes.number,
    swipeStartThreshold: PropTypes.number,
    threshold: PropTypes.number
  };

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  function _defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      Object.defineProperty(target, descriptor.key, descriptor);
    }
  }

  function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
  }

  function _defineProperty(obj, key, value) {
    if (key in obj) {
      Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
      });
    } else {
      obj[key] = value;
    }

    return obj;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function");
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        writable: true,
        configurable: true
      }
    });
    if (superClass) _setPrototypeOf(subClass, superClass);
  }

  function _getPrototypeOf(o) {
    _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
      return o.__proto__ || Object.getPrototypeOf(o);
    };
    return _getPrototypeOf(o);
  }

  function _setPrototypeOf(o, p) {
    _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
      o.__proto__ = p;
      return o;
    };

    return _setPrototypeOf(o, p);
  }

  function _assertThisInitialized(self) {
    if (self === void 0) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return self;
  }

  function _possibleConstructorReturn(self, call) {
    if (call && (typeof call === "object" || typeof call === "function")) {
      return call;
    }

    return _assertThisInitialized(self);
  }

  var styles$1 = {"swipeableListItem":"SwipeableListItem_swipeableListItem__3xnuO","contentRight":"SwipeableListItem_contentRight__Kox6B","contentLeft":"SwipeableListItem_contentLeft__3YoUt SwipeableListItem_contentRight__Kox6B","return":"SwipeableListItem_return__sR94A","contentLeftReturn":"SwipeableListItem_contentLeftReturn__1Ds5a SwipeableListItem_contentLeft__3YoUt SwipeableListItem_contentRight__Kox6B SwipeableListItem_return__sR94A","contentRightReturn":"SwipeableListItem_contentRightReturn__1YVPF SwipeableListItem_contentRight__Kox6B SwipeableListItem_return__sR94A","content":"SwipeableListItem_content__1k1mn","contentReturn":"SwipeableListItem_contentReturn__2_W7p SwipeableListItem_content__1k1mn"};

  var SwipeActionPropType = PropTypes.shape({
    action: PropTypes.func.isRequired,
    content: PropTypes.node.isRequired
  });
  var DragDirection = {
    UP: 1,
    DOWN: 2,
    LEFT: 3,
    RIGHT: 4,
    UNKNOWN: 5
  };
  var FPS_INTERVAL = 1000 / 60;

  var SwipeableListItem =
  /*#__PURE__*/
  function (_PureComponent) {
    _inherits(SwipeableListItem, _PureComponent);

    function SwipeableListItem(props) {
      var _this;

      _classCallCheck(this, SwipeableListItem);

      _this = _possibleConstructorReturn(this, _getPrototypeOf(SwipeableListItem).call(this, props));

      _defineProperty(_assertThisInitialized(_this), "resetState", function () {
        _this.dragStartPoint = {
          x: -1,
          y: -1
        };
        _this.dragDirection = DragDirection.UNKNOWN;
        _this.left = 0;
      });

      _defineProperty(_assertThisInitialized(_this), "handleDragStartMouse", function (event) {
        window.addEventListener('mouseup', _this.handleDragEndMouse);
        window.addEventListener('mousemove', _this.handleMouseMove);

        _this.wrapper.addEventListener('mouseup', _this.handleDragEndMouse);

        _this.wrapper.addEventListener('mousemove', _this.handleMouseMove);

        _this.handleDragStart(event);
      });

      _defineProperty(_assertThisInitialized(_this), "handleDragStartTouch", function (event) {
        window.addEventListener('touchend', _this.handleDragEndTouch);
        var touch = event.targetTouches[0];

        _this.handleDragStart(touch);
      });

      _defineProperty(_assertThisInitialized(_this), "handleDragStart", function (_ref) {
        var clientX = _ref.clientX,
            clientY = _ref.clientY;

        _this.resetState();

        _this.dragStartPoint = {
          x: clientX,
          y: clientY
        };
        _this.listElement.className = styles$1.content;

        if (_this.contentLeft) {
          _this.contentLeft.className = styles$1.contentLeft;
        }

        if (_this.contentRight) {
          _this.contentRight.className = styles$1.contentRight;
        }

        _this.startTime = Date.now();
        requestAnimationFrame(_this.updatePosition);
      });

      _defineProperty(_assertThisInitialized(_this), "handleTouchMove", function (event) {
        return _this.handleMove(event, event.targetTouches[0]);
      });

      _defineProperty(_assertThisInitialized(_this), "handleMouseMove", function (event) {
        return _this.handleMove(event, event);
      });

      _defineProperty(_assertThisInitialized(_this), "handleMove", function (event, _ref2) {
        var clientX = _ref2.clientX,
            clientY = _ref2.clientY;

        if (_this.dragStartedWithinItem()) {
          _this.setDragDirection(clientX, clientY);

          if (!event.cancelable) {
            return;
          }

          if (_this.isSwiping()) {
            event.stopPropagation();
            event.preventDefault();
            var delta = clientX - _this.dragStartPoint.x;

            if (_this.shouldMoveItem(delta)) {
              _this.left = delta;
            }
          }
        }
      });

      _defineProperty(_assertThisInitialized(_this), "handleDragEndMouse", function () {
        window.removeEventListener('mouseup', _this.handleDragEndMouse);
        window.removeEventListener('mousemove', _this.handleMouseMove);

        _this.wrapper.removeEventListener('mouseup', _this.handleDragEndMouse);

        _this.wrapper.removeEventListener('mousemove', _this.handleMouseMove);

        _this.handleDragEnd();
      });

      _defineProperty(_assertThisInitialized(_this), "handleDragEndTouch", function () {
        window.removeEventListener('touchend', _this.handleDragEndTouch);

        _this.handleDragEnd();
      });

      _defineProperty(_assertThisInitialized(_this), "handleDragEnd", function () {
        if (_this.isSwiping()) {
          var threshold = _this.props.threshold || 0.5;

          if (_this.left < _this.listElement.offsetWidth * threshold * -1) {
            _this.handleSwipedLeft();
          } else if (_this.left > _this.listElement.offsetWidth * threshold) {
            _this.handleSwipedRight();
          }
        }

        _this.resetState();

        _this.listElement.className = styles$1.contentReturn;
        _this.listElement.style.transform = "translateX(".concat(_this.left, "px)"); // hide backgrounds

        if (_this.contentLeft) {
          _this.contentLeft.style.opacity = 0;
          _this.contentLeft.className = styles$1.contentLeftReturn;
        }

        if (_this.contentRight) {
          _this.contentRight.style.opacity = 0;
          _this.contentRight.className = styles$1.contentRightReturn;
        }
      });

      _defineProperty(_assertThisInitialized(_this), "shouldMoveItem", function (delta) {
        var _this$props = _this.props,
            _this$props$swipeLeft = _this$props.swipeLeft;
        _this$props$swipeLeft = _this$props$swipeLeft === void 0 ? {} : _this$props$swipeLeft;
        var contentLeft = _this$props$swipeLeft.content,
            _this$props$swipeRigh = _this$props.swipeRight;
        _this$props$swipeRigh = _this$props$swipeRigh === void 0 ? {} : _this$props$swipeRigh;
        var contentRight = _this$props$swipeRigh.content,
            blockSwipe = _this$props.blockSwipe;
        var swipingLeft = delta < 0;
        var swipingRight = delta > 0;
        return !blockSwipe && (swipingLeft && contentLeft || swipingRight && contentRight);
      });

      _defineProperty(_assertThisInitialized(_this), "dragStartedWithinItem", function () {
        var _this$dragStartPoint = _this.dragStartPoint,
            x = _this$dragStartPoint.x,
            y = _this$dragStartPoint.y;
        return x !== -1 && y !== -1;
      });

      _defineProperty(_assertThisInitialized(_this), "setDragDirection", function (x, y) {
        if (_this.dragDirection === DragDirection.UNKNOWN) {
          var _this$dragStartPoint2 = _this.dragStartPoint,
              startX = _this$dragStartPoint2.x,
              startY = _this$dragStartPoint2.y;
          var horizontalDistance = Math.abs(x - startX);
          var verticalDistance = Math.abs(y - startY);

          if (horizontalDistance <= _this.dragHorizontalDirectionThreshold && verticalDistance <= _this.dragVerticalDirectionThreshold) {
            return;
          }

          var angle = Math.atan2(y - startY, x - startX);
          var octant = Math.round(8 * angle / (2 * Math.PI) + 8) % 8;

          switch (octant) {
            case 0:
              if (horizontalDistance > _this.dragHorizontalDirectionThreshold) {
                _this.dragDirection = DragDirection.RIGHT;
              }

              break;

            case 1:
            case 2:
            case 3:
              if (verticalDistance > _this.dragVerticalDirectionThreshold) {
                _this.dragDirection = DragDirection.DOWN;
              }

              break;

            case 4:
              if (horizontalDistance > _this.dragHorizontalDirectionThreshold) {
                _this.dragDirection = DragDirection.LEFT;
              }

              break;

            case 5:
            case 6:
            case 7:
              if (verticalDistance > _this.dragVerticalDirectionThreshold) {
                _this.dragDirection = DragDirection.UP;
              }

              break;
          }
        }
      });

      _defineProperty(_assertThisInitialized(_this), "isSwiping", function () {
        return _this.dragStartedWithinItem() && (_this.dragDirection === DragDirection.LEFT || _this.dragDirection === DragDirection.RIGHT);
      });

      _defineProperty(_assertThisInitialized(_this), "updatePosition", function () {
        var blockSwipe = _this.props.blockSwipe;

        if (_this.dragStartedWithinItem() && !blockSwipe) {
          requestAnimationFrame(_this.updatePosition);
        }

        var now = Date.now();
        var elapsed = now - _this.startTime;

        if (elapsed > FPS_INTERVAL && _this.isSwiping()) {
          var contentToShow = _this.left < 0 ? _this.contentLeft : _this.contentRight;

          if (!contentToShow) {
            return;
          }

          _this.listElement.style.transform = "translateX(".concat(_this.left, "px)");
          var opacity = (Math.abs(_this.left) / 100).toFixed(2);

          if (opacity < 1 && opacity.toString() !== contentToShow.style.opacity) {
            contentToShow.style.opacity = opacity.toString();
            var contentToHide = _this.left < 0 ? _this.contentRight : _this.contentLeft;

            if (contentToHide) {
              contentToHide.style.opacity = '0';
            }
          }

          if (opacity >= 1) {
            contentToShow.style.opacity = '1';
          }

          _this.startTime = Date.now();
        }
      });

      _defineProperty(_assertThisInitialized(_this), "handleSwipedLeft", function () {
        var _this$props$swipeLeft2 = _this.props.swipeLeft;
        _this$props$swipeLeft2 = _this$props$swipeLeft2 === void 0 ? {} : _this$props$swipeLeft2;
        var action = _this$props$swipeLeft2.action;

        if (action) {
          action();
        }
      });

      _defineProperty(_assertThisInitialized(_this), "handleSwipedRight", function () {
        var _this$props$swipeRigh2 = _this.props.swipeRight;
        _this$props$swipeRigh2 = _this$props$swipeRigh2 === void 0 ? {} : _this$props$swipeRigh2;
        var action = _this$props$swipeRigh2.action;

        if (action) {
          action();
        }
      });

      _defineProperty(_assertThisInitialized(_this), "bindContentLeft", function (ref) {
        return _this.contentLeft = ref;
      });

      _defineProperty(_assertThisInitialized(_this), "bindContentRight", function (ref) {
        return _this.contentRight = ref;
      });

      _defineProperty(_assertThisInitialized(_this), "bindListElement", function (ref) {
        return _this.listElement = ref;
      });

      _defineProperty(_assertThisInitialized(_this), "bindWrapper", function (ref) {
        return _this.wrapper = ref;
      });

      _this.contentLeft = null;
      _this.contentRight = null;
      _this.listElement = null;
      _this.wrapper = null;
      _this.startTime = null;

      _this.resetState();

      return _this;
    }

    _createClass(SwipeableListItem, [{
      key: "componentDidMount",
      value: function componentDidMount() {
        this.wrapper.addEventListener('mousedown', this.handleDragStartMouse);
        this.wrapper.addEventListener('touchstart', this.handleDragStartTouch);
        this.wrapper.addEventListener('touchend', this.handleDragEndTouch);
        this.wrapper.addEventListener('touchmove', this.handleTouchMove, {
          capture: true,
          passive: false
        });
      }
    }, {
      key: "componentWillUnmount",
      value: function componentWillUnmount() {
        this.wrapper.removeEventListener('mousedown', this.handleDragStartMouse);
        this.wrapper.removeEventListener('touchstart', this.handleDragStartTouch);
        this.wrapper.removeEventListener('touchend', this.handleDragEndTouch);
        this.wrapper.removeEventListener('touchmove', this.handleTouchMove, {
          capture: true,
          passive: false
        });
      }
    }, {
      key: "render",
      value: function render() {
        var _this$props2 = this.props,
            children = _this$props2.children,
            swipeLeft = _this$props2.swipeLeft,
            swipeRight = _this$props2.swipeRight;
        return React__default.createElement("div", {
          className: styles$1.swipeableListItem,
          ref: this.bindWrapper
        }, swipeLeft && React__default.createElement("div", {
          ref: this.bindContentLeft,
          className: styles$1.contentLeft,
          "data-testid": "swipe-left-content"
        }, swipeLeft.content), swipeRight && React__default.createElement("div", {
          ref: this.bindContentRight,
          className: styles$1.contentRight,
          "data-testid": "swipe-right-content"
        }, swipeRight.content), React__default.createElement("div", {
          ref: this.bindListElement,
          className: styles$1.content,
          "data-testid": "content"
        }, children));
      }
    }, {
      key: "dragHorizontalDirectionThreshold",
      get: function get() {
        return this.props.swipeStartThreshold || 10;
      }
    }, {
      key: "dragVerticalDirectionThreshold",
      get: function get() {
        return this.props.scrollStartThreshold || 10;
      }
    }]);

    return SwipeableListItem;
  }(React.PureComponent);

  SwipeableListItem.propTypes = {
    blockSwipe: PropTypes.bool,
    children: PropTypes.node.isRequired,
    swipeLeft: SwipeActionPropType,
    swipeRight: SwipeActionPropType,
    scrollStartThreshold: PropTypes.number,
    swipeStartThreshold: PropTypes.number,
    threshold: PropTypes.number
  };

  exports.SwipeableList = SwipeableList;
  exports.SwipeableListItem = SwipeableListItem;

  Object.defineProperty(exports, '__esModule', { value: true });

})));
